paymentsheet/src/main/java/com/stripe/android/customersheet/data/CustomerSessionPaymentMethodDataSource.kt

```diff
@@ -1,30 +1,72 @@
 package com.stripe.android.customersheet.data
 
+import com.stripe.android.core.injection.IOContext
 import com.stripe.android.model.PaymentMethod
 import com.stripe.android.model.PaymentMethodUpdateParams
+import com.stripe.android.payments.core.analytics.ErrorReporter
+import com.stripe.android.paymentsheet.repositories.CustomerRepository
+import kotlinx.coroutines.withContext
 import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
 
-internal class CustomerSessionPaymentMethodDataSource @Inject constructor() : CustomerSheetPaymentMethodDataSource {
+internal class CustomerSessionPaymentMethodDataSource @Inject constructor(
+    private val elementsSessionManager: CustomerSessionElementsSessionManager,
+    private val customerRepository: CustomerRepository,
+    private val errorReporter: ErrorReporter,
+    @IOContext private val workContext: CoroutineContext,
+) : CustomerSheetPaymentMethodDataSource {
     override suspend fun retrievePaymentMethods(): CustomerSheetDataResult<List<PaymentMethod>> {
-        throw NotImplementedError("Not implemented yet!")
+        return withContext(workContext) {
+            elementsSessionManager.fetchElementsSession().mapCatching { elementsSession ->
+                elementsSession.customer?.paymentMethods
+                    ?: throw IllegalStateException(
+                        "`CustomerSession` should contain payment methods in `elements/session` response!"
+                    )
```

Comment:
Going to follow-up with a change that handles this internally in `ElementsSessionManager`. `paymentMethods` field is not null but `customer` unexpectedly can.

Reply:
Done here: https://github.com/stripe/stripe-android/pull/9373